[DDP] queue autograd cb for static graph in first param's autograd hook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93105
* #93125

Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

rohan-varma committed:
[DDP] Use pre backward hook for delay allreduce

Per title

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use pre backward hook for delay allreduce"


Per title

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1155,6 +1172,13 @@ def forward(self, *inputs, **kwargs):
                 )
 
+            if self.static_graph and self.num_iterations == 1:
+                self._bwd_hook_handle = self.module.register_full_backward_pre_hook(
```
If we register it directly on `DDP` i.e. `self`, the firing order is off.

rohan-varma committed:
Update on "[DDP] Use pre backward hook for delay allreduce"


Switches use of _DDPSink to pre-backward hook for static_graph. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before.
Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Use pre backward hook for delay allreduce"


Switches use of _DDPSink to pre-backward hook for static_graph. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before.
Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] queue autograd cb for static graph in first param's autograd hook"


Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] queue autograd cb for static graph in first param's autograd hook"


Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -8902,7 +8902,7 @@ class MyDataClass:
                 def __init__(self):
                     super().__init__()
-                    self.m = torch.nn.Linear(10, 10)
+                    self.m = torch.nn.Linear(10, 10, bias=False)
```
this change is mainly to test that the autograd cb indeed gets enqueued if there is only one param that gets grad, as expected.

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1040,9 +1040,12 @@ def setUp(self):
-        store = self._get_store()
-        c10d.init_process_group("nccl", store=store, rank=self.rank, world_size=self.world_size)
-        return c10d.distributed_c10d._get_default_group()
+        try:
```
Yep, without this, we hit an error about trying to init default pg twice.

rohan-varma committed:
Update on "[DDP] queue autograd cb for static graph in first param's autograd hook"


Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] queue autograd cb for static graph in first param's autograd hook"


Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -631,7 +631,11 @@ void Reducer::autograd_hook(size_t index) {
   }
 
-  if (static_graph_first_iteration()) {
+  if (static_graph_first_iteration() && !static_graph_first_iteration_cb_enqueued_) {
```
will this still work if the first parameter (or all parameters) are not used in one rank? Or do we never expect delay_all_reduce() to be combined with unused parameters?

If unused parameters is a concern here, will it be more generic if we do `_find_tensors` on inputs or outputs and call queue callback there?

rohan-varma replied:
This works if the first param is not used on a particular rank, and the unittest that has a model with a single param confirms this.

We'd have to investigate a bit more in the case all params are not used on a particular rank. We can try the following case:

def forward(x):
    if rank == 0:
        x = self.lin(x)
    return x

and check whether the current and new methods of delaying allreduce work or not.

rohan-varma committed:
Update on "[DDP] queue autograd cb for static graph in first param's autograd hook"


Switches use of _DDPSink to queueing autograd callback in the first param's autograd hook. 

This enables:
1. Long term we can get rid of _DDPSink once we do this for unused parameters
2. Better support custom output types such as dataclass from which tensors possibly could not be extracted from before, as shown by unittest. 

Differential Revision: [D42789574](https://our.internmc.facebook.com/intern/diff/D42789574/)

[ghstack-poisoned]
